File: cses/02_sorting_searching/2216_collecting-numbers/solution.py
import sys


def main():
    # Đọc toàn bộ input một lần cho nhanh
    data = sys.stdin.buffer.read().split()
    if not data:
        return
    n = int(data[0])
    a = data[1:1 + n]

    # pos[v] = vị trí (index) của giá trị v trong mảng
    pos = [0] * (n + 1)
    for i in range(n):
        pos[int(a[i])] = i

    # Mỗi lần pos[v] < pos[v-1] là một "đứt gãy" cần thêm một vòng mới;
    # số vòng = 1 + số đứt gãy
    rounds = 1
    for v in range(2, n + 1):
        if pos[v] < pos[v - 1]:
            rounds += 1

    sys.stdout.write(str(rounds) + "\n")


main()

File: cses/02_sorting_searching/2216_collecting-numbers/solution.cpp
#include <iostream>
#include <vector>

using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;

    // position[v] = chỉ số (index) nơi giá trị v xuất hiện trong mảng
    vector<int> position(n + 1);
    for (int i = 0; i < n; ++i) {
        int value;
        cin >> value;
        position[value] = i;
    }

    // Mỗi lần position[v] < position[v-1] là một "đứt gãy" cần thêm một vòng mới;
    // số vòng = 1 + số đứt gãy
    int rounds = 1;
    for (int value = 2; value <= n; ++value) {
        if (position[value] < position[value - 1]) {
            ++rounds;
        }
    }

    cout << rounds << '\n';
    return 0;
}
